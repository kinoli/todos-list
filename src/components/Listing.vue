<template lang="html">

  <section class="listing">
    <v-layout row>
      <v-flex xs12 sm6 offset-sm3>
        <v-card>
          <v-toolbar color="teal" dark>
            <v-toolbar-title class="text-xs-center">Todos</v-toolbar-title>

            <v-spacer />

            <v-switch
              id="completed-switch"
              label="Completed?"
              v-model="completed"
              @change="onFilterOrSortChanged"
              color="yellow"
              class="completed-switch mr-4 mt-2"
            ></v-switch>

            <v-text-field
              id="search-input"
              v-model="searchTerm"
              v-on:input="onFilterOrSortChanged"
              class="mt-2"
              flat
              label="Filter"
              prepend-inner-icon="search"
              solo-inverted
            ></v-text-field>
          </v-toolbar>

          <v-list two-line>
            <div v-for="(todo, index) in todos" :key="todo.id">
              <ListingRow class="l" :todo="todo" />
              <v-divider v-if="index + 1 < todos.length" :key="`divider-${index}`"></v-divider>
            </div>
          </v-list>
        </v-card>
      </v-flex>
    </v-layout>
  </section>

</template>

<script lang="js">
import ListingRow from './ListingRow'

export default {
  name: 'listing',
  components: { ListingRow },
  mounted () {
    this.searchTerm = this.$store.getters.getSearchTerm
    this.completed = this.$store.getters.getCompletedToggle
    this.originalTodos = this.$store.getters.getTodosWithUsers
    this.todos = this.originalTodos

    const modifiedTodos = this.getSortCompleted()
    this.doSearch(modifiedTodos)
  },
  data: () => ({
    loaded: false,
    todos: [],
    users: [],
    originalTodos: [],
    searchTerm: '',
    completed: false
  }),
  methods: {
    onFilterOrSortChanged () {
      this.$store.dispatch('SET_COMPLETED_TOGGLE', this.completed)
      this.$store.dispatch('SET_SEARCH_TERM', this.searchTerm)

      const modifiedTodos = this.getSortCompleted()
      this.doSearch(modifiedTodos)
    },
    doSearch (modifiedTodos) {
      if (!this.searchTerm) {
        this.todos = modifiedTodos
      }
      this.todos = modifiedTodos.filter(t => {
        const lowerTerm = this.searchTerm.toLowerCase()
        if (
          t.title.toLowerCase().includes(lowerTerm) ||
          t.user.name.toLowerCase().includes(lowerTerm) ||
          t.user.username.toLowerCase().includes(lowerTerm) ||
          t.user.company.name.toLowerCase().includes(lowerTerm)
        ) {
          return true
        }
      })
    },
    getSortCompleted () {
      if (!this.completed) {
        return this.originalTodos
      }
      return this.originalTodos.filter(t => t.completed === this.completed)
    }
  }
}
</script>

<style lang="scss">
  .listing {
    .completed-switch {
      justify-content: flex-end;
      flex-grow: 0;

      .v-input__slot {
        margin-bottom: 0;
      }
    }
  }
</style>
