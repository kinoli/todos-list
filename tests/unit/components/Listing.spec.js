import Vue from 'vue'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Listing from '@/components/Listing.vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(Vuex)

const mockState = {
  todos: [
    {
      userId: 1,
      id: 1,
      title: 'one two three',
      completed: false
    },
    {
      userId: 1,
      id: 2,
      title: 'two three four',
      completed: true
    },
    {
      userId: 2,
      id: 3,
      title: 'three four five',
      completed: false
    }
  ],
  users: [
    {
      id: 1,
      name: 'Leanne Graham',
      username: 'policy',
      company: {
        name: 'Foo'
      }
    },
    {
      id: 2,
      name: 'Ervin Howell',
      username: 'fly',
      company: {
        name: 'Bar'
      }
    }
  ],
  tab: null,
  searchTerm: '',
  completedToggle: false
}

const mockTodosWithUsers = mockState.todos.map(t => {
  const user = mockState.users && mockState.users.filter(u => u.id === t.userId)
  return { ...t, user: user && user.length && user[0] }
})

// this is somewhat of a hack but since I just want to shallow test and
// not have to load an additional library, it works fine here
const mockInputChange = (searchTerm, completed, lengthShouldBe) => {
  wrapper.vm.searchTerm = searchTerm
  wrapper.vm.completed = completed
  wrapper.vm.onFilterOrSortChanged()
  expect(wrapper.findAll('listingrow-stub').length).toEqual(lengthShouldBe)
}

let wrapper, actions, getters, store
beforeEach(() => {
  actions = {
    SET_COMPLETED_TOGGLE: jest.fn(),
    SET_SEARCH_TERM: jest.fn()
  }
  getters = {
    getSearchTerm: () => (mockState.searchTerm),
    getCompletedToggle: () => (mockState.completedToggle),
    getTodosWithUsers: () => (mockTodosWithUsers)
  }
  store = new Vuex.Store({
    state: mockState,
    actions,
    getters
  })
  wrapper = shallowMount(Listing, { store, localVue })
})

describe('Listing.vue', () => {
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.searchTerm).toEqual(mockState.searchTerm)
    expect(wrapper.vm.todos).toEqual(mockTodosWithUsers)
  })

  it('stores inputs when a change is made', () => {
    mockInputChange('one', false, 1)
    expect(actions.SET_SEARCH_TERM).toHaveBeenCalled()
    expect(actions.SET_COMPLETED_TOGGLE).toHaveBeenCalled()
  })

  it('should filter list accurately with input change', () => {
    mockInputChange('one', false, 1)
    mockInputChange('One', false, 1)
    mockInputChange('two', false, 2)
    mockInputChange('Three', false, 3)
  })

  it('should sort list by showing only completed todos', () => {
    mockInputChange('', false, 3)
    mockInputChange('', true, 1)
  })

  it('should sort and filter list accurately', () => {
    mockInputChange('one', true, 0)
    mockInputChange('two', false, 2)
    mockInputChange('two', true, 1)
    mockInputChange('three', true, 1)
    mockInputChange('five', false, 1)
    mockInputChange('five', true, 0)
  })
})
