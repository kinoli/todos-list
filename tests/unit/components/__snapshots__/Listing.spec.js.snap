// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing.vue renders correctly 1`] = `
<section
  class="listing"
>
  <v-layout-stub
    row=""
    tag="div"
  >
    <v-flex-stub
      offset-sm3=""
      sm6=""
      tag="div"
      xs12=""
    >
      <v-card-stub>
        <v-toolbar-stub
          color="teal"
          dark=""
          scrollthreshold="300"
        >
          <v-toolbar-title-stub
            class="text-xs-center"
          >
            Todos
          </v-toolbar-title-stub>
           
          <v-spacer-stub />
           
          <v-switch-stub
            class="completed-switch mr-4 mt-2"
            color="yellow"
            id="completed-switch"
            label="Completed?"
          />
           
          <v-text-field-stub
            class="mt-2"
            flat=""
            id="search-input"
            label="Filter"
            prepend-inner-icon="search"
            solo-inverted=""
          />
        </v-toolbar-stub>
         
        <v-list-stub
          twoline="true"
        >
          <div>
            <listingrow-stub
              class="l"
              todo="[object Object]"
            />
             
            <v-divider-stub />
          </div>
          <div>
            <listingrow-stub
              class="l"
              todo="[object Object]"
            />
             
            <v-divider-stub />
          </div>
          <div>
            <listingrow-stub
              class="l"
              todo="[object Object]"
            />
             
            <!---->
          </div>
        </v-list-stub>
      </v-card-stub>
    </v-flex-stub>
  </v-layout-stub>
</section>
`;
